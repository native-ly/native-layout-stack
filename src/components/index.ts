export { Stack } from './Stack'
export { Spacer } from './Spacer'
