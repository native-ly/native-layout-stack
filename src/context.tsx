import React from 'react'

import type { Padding } from './types/Padding'

export interface Props {
  readonly debug: boolean
  readonly spaces?: number
  readonly padding?: Padding
}

export const LayoutContext = React.createContext<Props>({
  debug: false,
  spaces: undefined,
  padding: undefined,
})

LayoutContext.displayName = 'LayoutContext'

export const LayoutContextProvider: React.FC<Props> = ({
  children,
  ...props
}) => <LayoutContext.Provider value={props}>{children}</LayoutContext.Provider>
