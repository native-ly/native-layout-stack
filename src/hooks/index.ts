export { useStack } from './useStack'
export { useDebugStyle } from './useDebugStyle'
export { useSpacing } from './useSpacing'
export { useSpacingHelpers } from './useSpacingHelpers'
