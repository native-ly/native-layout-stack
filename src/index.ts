export { LayoutProvider } from './context'
export * from './components'
