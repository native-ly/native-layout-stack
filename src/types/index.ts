export type { Padding } from './Padding'
export type { Spaces } from './Spaces'
