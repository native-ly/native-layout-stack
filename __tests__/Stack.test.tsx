import React from 'react'
import { Text } from 'react-native'
import { render } from '@testing-library/react-native'

import { Stack, LayoutProvider } from '../src'

const BareComponent = () => (
  <Stack>
    <Text>Hello</Text>

    <Text>World</Text>
  </Stack>
)

describe('Stack', () => {
  it('should render component without context provider', () => {
    const { toJSON } = render(
      <Stack spaces={20} padding={10}>
        <Text>Hello</Text>

        <Text>World</Text>
      </Stack>
    )

    expect(toJSON()).toMatchSnapshot()
  })

  it('should render component with custom styles', () => {
    const { toJSON } = render(
      <Stack
        spaces={20}
        padding={10}
        style={{ backgroundColor: 'green', padding: 5 }}
      >
        <Text>Hello</Text>

        <Text>World</Text>
      </Stack>
    )

    expect(toJSON()).toMatchSnapshot()
  })

  it('should render component with context provider', () => {
    const { toJSON } = render(
      <LayoutProvider spaces={15} padding={30}>
        <BareComponent />
      </LayoutProvider>
    )

    expect(toJSON()).toMatchSnapshot()
  })

  it('should render component with debug mode enabled', () => {
    const { toJSON } = render(
      <LayoutProvider spaces={15} padding={30} debug>
        <BareComponent />
      </LayoutProvider>
    )

    expect(toJSON()).toMatchSnapshot()
  })

  it('should overwrite global options', () => {
    const Component = () => (
      <Stack spaces={12} padding={16}>
        <Text>Hello</Text>

        <Stack spaces={5}>
          <Text>World</Text>
        </Stack>
      </Stack>
    )

    const { toJSON } = render(
      <LayoutProvider spaces={15} padding={30}>
        <Component />
      </LayoutProvider>
    )

    expect(toJSON()).toMatchSnapshot()
  })

  it.each([
    12,
    [12],
    [12, 6],
    [5, 10, 15],
    [10, 20, 30, 40],
    '12',
    ['12'],
    ['12', '6'],
    ['5', '10', '15'],
    ['10', '20', '30', '40'],
    ['12', 6],
    [5, '10', '15'],
    ['10', 20, 30, '40'],
  ])('', (padding: any) => {
    const { toJSON } = render(
      <Stack spaces={20} padding={padding}>
        <Text>Hello</Text>

        <Text>World</Text>
      </Stack>
    )

    expect(toJSON()).toMatchSnapshot()
  })

  // TODO update test
  it.skip('sholud throw an error', () => {
    render(
      <Stack padding={[1, 2, 3, 4, 5] as any}>
        <Text>Hello</Text>

        <Text>World</Text>
      </Stack>
    )
  })

  // TODO update test
  it.skip('sholud throw an error', () => {
    render(
      <Stack spaces={[] as any}>
        <Text>Hello</Text>

        <Text>World</Text>
      </Stack>
    )
  })

  it('should render component with horizontal layout', () => {
    const { toJSON } = render(
      <Stack spaces={20} padding={10} style={{ flexDirection: 'row' }}>
        <Text>Hello</Text>

        <Text>World</Text>
      </Stack>
    )

    expect(toJSON()).toMatchSnapshot()
  })
})
