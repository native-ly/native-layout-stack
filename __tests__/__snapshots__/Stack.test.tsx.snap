// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stack  1`] = `
<View
  style={
    Object {
      "padding": 12,
    }
  }
>
  <Text>
    Hello
  </Text>
  <View
    style={
      Object {
        "minHeight": 20,
        "minWidth": 20,
      }
    }
  />
  <Text>
    World
  </Text>
</View>
`;

exports[`Stack  2`] = `
<View
  style={
    Object {
      "padding": 12,
    }
  }
>
  <Text>
    Hello
  </Text>
  <View
    style={
      Object {
        "minHeight": 20,
        "minWidth": 20,
      }
    }
  />
  <Text>
    World
  </Text>
</View>
`;

exports[`Stack  3`] = `
<View
  style={
    Object {
      "paddingHorizontal": 6,
      "paddingVertical": 12,
    }
  }
>
  <Text>
    Hello
  </Text>
  <View
    style={
      Object {
        "minHeight": 20,
        "minWidth": 20,
      }
    }
  />
  <Text>
    World
  </Text>
</View>
`;

exports[`Stack  4`] = `
<View
  style={
    Object {
      "paddingBottom": 15,
      "paddingHorizontal": 10,
      "paddingTop": 5,
    }
  }
>
  <Text>
    Hello
  </Text>
  <View
    style={
      Object {
        "minHeight": 20,
        "minWidth": 20,
      }
    }
  />
  <Text>
    World
  </Text>
</View>
`;

exports[`Stack  5`] = `
<View
  style={
    Object {
      "paddingBottom": 30,
      "paddingLeft": 40,
      "paddingRight": 20,
      "paddingTop": 10,
    }
  }
>
  <Text>
    Hello
  </Text>
  <View
    style={
      Object {
        "minHeight": 20,
        "minWidth": 20,
      }
    }
  />
  <Text>
    World
  </Text>
</View>
`;

exports[`Stack  6`] = `
<View
  style={
    Object {
      "padding": "12",
    }
  }
>
  <Text>
    Hello
  </Text>
  <View
    style={
      Object {
        "minHeight": 20,
        "minWidth": 20,
      }
    }
  />
  <Text>
    World
  </Text>
</View>
`;

exports[`Stack  7`] = `
<View
  style={
    Object {
      "padding": "12",
    }
  }
>
  <Text>
    Hello
  </Text>
  <View
    style={
      Object {
        "minHeight": 20,
        "minWidth": 20,
      }
    }
  />
  <Text>
    World
  </Text>
</View>
`;

exports[`Stack  8`] = `
<View
  style={
    Object {
      "paddingHorizontal": "6",
      "paddingVertical": "12",
    }
  }
>
  <Text>
    Hello
  </Text>
  <View
    style={
      Object {
        "minHeight": 20,
        "minWidth": 20,
      }
    }
  />
  <Text>
    World
  </Text>
</View>
`;

exports[`Stack  9`] = `
<View
  style={
    Object {
      "paddingBottom": "15",
      "paddingHorizontal": "10",
      "paddingTop": "5",
    }
  }
>
  <Text>
    Hello
  </Text>
  <View
    style={
      Object {
        "minHeight": 20,
        "minWidth": 20,
      }
    }
  />
  <Text>
    World
  </Text>
</View>
`;

exports[`Stack  10`] = `
<View
  style={
    Object {
      "paddingBottom": "30",
      "paddingLeft": "40",
      "paddingRight": "20",
      "paddingTop": "10",
    }
  }
>
  <Text>
    Hello
  </Text>
  <View
    style={
      Object {
        "minHeight": 20,
        "minWidth": 20,
      }
    }
  />
  <Text>
    World
  </Text>
</View>
`;

exports[`Stack  11`] = `
<View
  style={
    Object {
      "paddingHorizontal": 6,
      "paddingVertical": "12",
    }
  }
>
  <Text>
    Hello
  </Text>
  <View
    style={
      Object {
        "minHeight": 20,
        "minWidth": 20,
      }
    }
  />
  <Text>
    World
  </Text>
</View>
`;

exports[`Stack  12`] = `
<View
  style={
    Object {
      "paddingBottom": "15",
      "paddingHorizontal": "10",
      "paddingTop": 5,
    }
  }
>
  <Text>
    Hello
  </Text>
  <View
    style={
      Object {
        "minHeight": 20,
        "minWidth": 20,
      }
    }
  />
  <Text>
    World
  </Text>
</View>
`;

exports[`Stack  13`] = `
<View
  style={
    Object {
      "paddingBottom": 30,
      "paddingLeft": "40",
      "paddingRight": 20,
      "paddingTop": "10",
    }
  }
>
  <Text>
    Hello
  </Text>
  <View
    style={
      Object {
        "minHeight": 20,
        "minWidth": 20,
      }
    }
  />
  <Text>
    World
  </Text>
</View>
`;

exports[`Stack should overwrite global options 1`] = `
<View
  style={
    Object {
      "padding": 16,
    }
  }
>
  <Text>
    Hello
  </Text>
  <View
    style={
      Object {
        "minHeight": 12,
        "minWidth": 12,
      }
    }
  />
  <View
    style={
      Object {
        "padding": 30,
      }
    }
  >
    <Text>
      World
    </Text>
  </View>
</View>
`;

exports[`Stack should render component with context provider 1`] = `
<View
  style={
    Object {
      "padding": 30,
    }
  }
>
  <Text>
    Hello
  </Text>
  <View
    style={
      Object {
        "minHeight": 15,
        "minWidth": 15,
      }
    }
  />
  <Text>
    World
  </Text>
</View>
`;

exports[`Stack should render component with custom styles 1`] = `
<View
  style={
    Object {
      "backgroundColor": "green",
      "padding": 10,
    }
  }
>
  <Text>
    Hello
  </Text>
  <View
    style={
      Object {
        "minHeight": 20,
        "minWidth": 20,
      }
    }
  />
  <Text>
    World
  </Text>
</View>
`;

exports[`Stack should render component with debug mode enabled 1`] = `
<View
  style={
    Object {
      "padding": 30,
    }
  }
>
  <Text>
    Hello
  </Text>
  <View
    style={
      Object {
        "backgroundColor": "#f0f",
        "minHeight": 15,
        "minWidth": 15,
      }
    }
  />
  <Text>
    World
  </Text>
</View>
`;

exports[`Stack should render component with horizontal layout 1`] = `
<View
  style={
    Object {
      "flexDirection": "row",
      "padding": 10,
    }
  }
>
  <Text>
    Hello
  </Text>
  <View
    style={
      Object {
        "minHeight": 20,
        "minWidth": 20,
      }
    }
  />
  <Text>
    World
  </Text>
</View>
`;

exports[`Stack should render component without context provider 1`] = `
<View
  style={
    Object {
      "padding": 10,
    }
  }
>
  <Text>
    Hello
  </Text>
  <View
    style={
      Object {
        "minHeight": 20,
        "minWidth": 20,
      }
    }
  />
  <Text>
    World
  </Text>
</View>
`;
